app/javascript/components/tree-view/hierarchical-tree-view.jsx

```diff
@@ -1,72 +1,161 @@
+
+const callBack = (nodeId, type, value, namespace) => ({
+  nodeId, type, value, namespace,
+});
+
+/**
+ * Bottom to top checkbox checking.
+*/
+const postProcessingCheck = tree => tree.map((n) => {
+  let node = { ...n };
+
+  if (node.nodes) {
+    node = { ...node, nodes: postProcessingCheck(node.nodes) };
+    const checkedNum = node.nodes.filter(node => node.state.checked).length;
+    if (node.nodes.length === checkedNum) {
+      node = { ...node, state: { ...node.state, checked: true } };
+    } else if (checkedNum > 0) {
+      node = { ...node, state: { ...node.state, checked: null } };
+    } else {
+      node = { ...node, state: { ...node.state, checked: false } };
+    }
+  }
+  return node;
+});
+
+/**
+ * Helper
+ * * Changes class to className.
+ */
+const convert = tree => tree.map((n) => {
+  let node = { ...n };
+  if (node.class) {
+    node = { ...node, classes: node.class };
+    delete node.class;
+  }
+
+  if (node.key) {
+    node = { ...node, attr: { key: node.key } };
+    delete node.key;
+  }
+
+  if (node.nodes) {
+    node = { ...node, nodes: convert(node.nodes) };
+  }
+
+  return node;
+});
 
 const HierarchicalTreeView = (props) => {
   const {
     tree_name,
     bs_tree,
     check_url,
+    click_url,
     oncheck,
+    onclick,
     checkboxes,
     allow_reselect,
+    post_check,
+    hierarchical_check,
     callBack,
   } = props;
 
-  const storeName = `tree_${tree_name}`;
-  const ConnectedNode = connect((store, ownProps) => ({ ...store[storeName][ownProps.nodeId] }))(Node);
-  const ReduxTree = connect(store => ({ data: { ...store[storeName] } }))(Tree);
+  const namespace = tree_name;
+  const ConnectedNode = connect((store, ownProps) => ({ ...store[namespace][ownProps.nodeId] }))(Node);
+  const ReduxTree = connect(store => ({ data: { ...store[namespace] } }))(Tree);
+
 
   /**
    * After the component mounts adds a tree specific reducer to the store
    */
   useEffect(() => {
-    ManageIQ.redux.addReducer({ [storeName]: defaultStore });
+    // Setting the URLs for the further actions:
+    ManageIQ.tree.checkUrl = check_url;
+    ManageIQ.tree.clickUrl = click_url;
+
+    ManageIQ.redux.addReducer({
+      [namespace]: combineReducers([
+        basicStore,
+        reducers(oncheck, onclick, tree_name),
+      ], namespace),
+    });
   }, []);
 
   /**
    * Populates the store from the prop by converting the supplied tree to
    * the correct format and then dispatching it to the store.
    */
   useEffect(() => {
-    callBack('', ActionTypes.ADD_NODES, convertData(bs_tree));
+    // FIXME - When the conversion wont be needed hopefuly in the future
+    const _tree = convert(JSON.parse(bs_tree));
+    const tree = (post_check && hierarchical_check) ? postProcessingCheck(_tree) : _tree;
+
+    callBack(null, ACTIONS.EMPTY_TREE, null, namespace);
+    callBack('', ACTIONS.ADD_NODES, convertData(tree), namespace);
```

Comment:
This is confusing, what's the difference between the behavior with `null` and `''`?

Reply:
Well, nothing anymore. Thanks for highlighting.